Fix double compile

stefnotch opened the pull request:
Fix #163

According to [this source](https://users.rust-lang.org/t/why-is-this-crate-built-twice-when-using-profile-test), dependencies of proc macros get compiled twice. Which can really hurt our compile times.

[`eval/prelude.rs`](https://github.com/wgsl-tooling-wg/wesl-rs/blob/64ae2373605af54be3a4baf251d673683da85fe9/crates/wesl/src/eval/prelude.rs#L20) makes use of the `quote_module!` proc macro. Which then causes the extra compilation of wgsl_parse. This PR simply moves the `wgsl_parse` dependency behind the feature flag, which avoids the hit in the default configuration. As soon as one turns `eval` on, one gets this terrible compile times hit.

I think we should
- Replace the use of the `quote_module!` macro in the prelude. I think we can absolutely get away with parsing the prelude at runtime instead of doing it at compile time.
- Consider removing the `quote_*` family of macros, since they do come with this compilation time footgun. naga doesn't provide anything comparable, so I suspect it's fine to just not have that feature. 
  - There could also be a possibility of not invoking `wgsl_parse` when running the proc macro, and instead just generating the code that would invoke `wgsl_parse` at runtime?

k2d222 commented:
thanks.
yes, there is no need for the quote functions in the prelude at the moment.

a future PR can focus on reverting usage of quote in the prelude, and at the same time we can think of what to do with this feature: keep it, move to separate crate or remove altogeter.

stefnotch commented:
@k2d222 I simplified the old example, and renamed it to `buildrs_wesl`.

While doing that, I also removed the usage of `wesl::quote_module!`. I'm not sure what to do with that feature for now.

And I replaced `let source = wesl::include_wesl!("main");` with the expanded form of `let source = include_str!(concat!(env!("OUT_DIR"), "/", "main", ".wgsl"));`
While I absolutely like the `include_wesl!` macro, it's not worth the massive compile time hit. It's only useful if I already have wesl-rs both as a normal dependency and as a build dependency. I wonder if there's a nicer solution?

k2d222 commented:
> And I replaced let source = wesl::include_wesl!("main"); with the expanded form of let source = include_str!(concat!(env!("OUT_DIR"), "/", "main", ".wgsl"));

this is awkward. I took this design idea from [lalrpop](https://lalrpop.github.io/lalrpop/tutorial/001_adding_lalrpop.html). They have two crates, `lalrpop` and `lalrpop_util`. The former for build-time and the latter contains only the `lalrpop_mod!` macro (plus a couple types) analogous to `include_wesl!`.

I think we can do the same. There might even be another advantage: adding a dependency at runtime pins the `wesl-rs` version during dependency resolution (all dependent packages must have a semver-compatible version of `wesl-rs`.

WDYT?
otherwise please fix the markdownlint complaint and the PR is ready ✨

stefnotch commented:
@k2d222 That does sound like a reasonable design.

As an alternative:
What I do in my projects is taking all shaders, compiling them to wgsl and *generating bindings*.
Then I bundle all that into a single Rust file. At this point, all I need to do is adding a single `include_str!(concat!(env!("OUT_DIR"), "/shaders.rs"));`. 

I also sometimes generate that file into the `src` folder, since rust-analyzer really struggles with including generated files from the build directory. Not sure why.

stefnotch commented:
I can fix the markdown lint next week, when I'm back on the machine where I made this PR. I'm not sure how I feel about those markdown lints, since I cannot reasonably use a markdown formatter extension. There are too many projects where auto-formatting their markdown files makes the maintainers upset.

I could have a markdown formatter extension just for the wesl-rs repository, but I'm one of the terrible people that avoid extra tools and dependencies if they can.

k2d222 commented:
Ok, I'll fix the lint and open a separate issue for include_wesl.
